src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/SpellCheck/SpellCheckService.cs

```diff
@@ -0,0 +1,137 @@
+    {
+        // We don't want to report spelling errors in script or style tags, so we avoid descending into them at all, which
+        // means we don't need complicated logic, and it performs a bit better. We assume any C# in them will still be reported
+        // by Roslyn.
+        // In an ideal world we wouldn't need this logic at all, as we would defer to the Html LSP server to provide spell checking
+        // but it doesn't currently support it. When that support is added, we can remove all of this but the RazorCommentBlockSyntax
+        // handling.
+        foreach (var node in syntaxTree.Root.DescendantNodes(static n => n is not MarkupElementSyntax { StartTag.Name.Content: "script" or "style" }))
+        {
+            if (node is RazorCommentBlockSyntax commentBlockSyntax)
+            {
+                ranges.Add(new((int)VSInternalSpellCheckableRangeKind.Comment, commentBlockSyntax.Comment.SpanStart, commentBlockSyntax.Comment.Span.Length));
+            }
+            else if (node is MarkupTextLiteralSyntax textLiteralSyntax)
+            {
+                // Attribute names are text literals, but we don't want to spell check them because either C# will,
+                // whether they're component attributes based on property names, or they come from tag helper attribute
+                // parameters as strings, or they're Html attributes which are not necessarily expected to be real words.
+                if (node.Parent is MarkupTagHelperAttributeSyntax or
+                    MarkupAttributeBlockSyntax or
+                    MarkupMinimizedAttributeBlockSyntax or
+                    MarkupTagHelperDirectiveAttributeSyntax or
+                    MarkupMinimizedTagHelperAttributeSyntax or
+                    MarkupMinimizedTagHelperDirectiveAttributeSyntax or
+                    MarkupMiscAttributeContentSyntax)
+                {
+                    continue;
+                }
+
+                // Text literals appear everywhere in Razor to hold newlines and indentation, so its worth saving the tokens
+                if (textLiteralSyntax.ContainsOnlyWhitespace())
+                {
+                    continue;
+                }
+
+                if (textLiteralSyntax.Span.Length == 0)
+                {
+                    continue;
+                }
+
+                ranges.Add(new((int)VSInternalSpellCheckableRangeKind.String, textLiteralSyntax.SpanStart, textLiteralSyntax.Span.Length));
+            }
+        }
+    }
+
+    private void AddCSharpSpellCheckRanges(ref PooledArrayBuilder<SpellCheckRange> ranges, ImmutableArray<SpellCheckRange> csharpRanges, RazorCodeDocument codeDocument)
+    {
+        var csharpDocument = codeDocument.GetCSharpDocument();
+
+        foreach (var report in csharpRanges)
+        {
+            var absoluteCSharpStartIndex = report.AbsoluteStartIndex;
+            var length = report.Length;
```

Comment:
Since `SpellCheckRange` is a record, consider using deconstruction to assign to local variables rather than grabbing them individually.

```suggestion
            var (kind, absoluteCSharpStartIndex, length) = report;
```

Reply:
I am not a fan of deconstructing this into what is essentially 3 arbitrary ints. I guess tests should fail if the order is wrong, but I still don't like it. Fortunately, your suggestion to use a `with` makes it less attractive anyway, as we don't need `kind` any more :)